src/cipher.rs

```diff
@@ -472,319 +486,651 @@ impl rustls_root_cert_store {
         ffi_panic_boundary! {
-            let client_cert_verifier_builder: &mut Option<AllowAnyAnonymousOrAuthenticatedClient> = try_mut_from_ptr!(builder);
+            let server_verifier_builder: &mut Option<ServerCertVerifierBuilder> = try_mut_from_ptr!(builder);
+            let server_verifier_builder = try_take!(server_verifier_builder);
 
-            let client_cert_verifier = match client_cert_verifier_builder.take() {
-                None => {
-                    return null() as *const _;
-                },
-                Some(x) => x,
+            let mut builder = WebPkiServerVerifier::builder(server_verifier_builder.roots)
+                .with_crls(server_verifier_builder.crls);
+            if matches!(server_verifier_builder.revocation_depth, RevocationCheckDepth::EndEntity) {
+                builder = builder.only_check_end_entity_revocation();
+            }
+            if matches!(server_verifier_builder.revocation_policy, UnknownStatusPolicy::Allow) {
+                builder = builder.allow_unknown_revocation_status();
+            }
+
+            let verifier = match builder.build() {
+                Ok(v) => v,
+                Err(e) => return error::map_verifier_builder_error(e),
             };
-            Arc::into_raw(client_cert_verifier.boxed()) as *const _
+
+            set_boxed_mut_ptr(verifier_out, verifier);
+
+            rustls_result::Ok
         }
     }
 
-    /// "Free" a verifier previously returned from
-    /// `rustls_allow_any_anonymous_or_authenticated_client_verifier_new`. Since
-    /// `rustls_allow_any_anonymous_or_authenticated_client_verifier`
-    /// is actually an atomically reference-counted pointer, extant `server_configs` may still
-    /// hold an internal reference to the Rust object. However, C code must
-    /// consider this pointer unusable after "free"ing it.
-    /// Calling with NULL is fine. Must not be called twice with the same value.
+    /// Free a `rustls_server_cert_verifier_builder` previously returned from
+    /// `rustls_server_cert_verifier_builder_new`. Calling with NULL is fine. Must not be
+    /// called twice with the same value.
     #[no_mangle]
-    pub extern "C" fn rustls_allow_any_anonymous_or_authenticated_client_verifier_free(
-        verifier: *const rustls_allow_any_anonymous_or_authenticated_client_verifier,
+    pub extern "C" fn rustls_web_pki_server_cert_verifier_builder_free(
+        builder: *mut rustls_web_pki_server_cert_verifier_builder,
     ) {
         ffi_panic_boundary! {
-            free_arc(verifier);
+            free_box(builder);
+        }
+    }
+}
+
+/// A built server certificate verifier that can be provided to a `rustls_client_config_builder`
+/// with `rustls_client_config_builder_set_server_verifier`.
+pub struct rustls_server_cert_verifier {
+    _private: [u8; 0],
+}
+
+impl Castable for rustls_server_cert_verifier {
+    type Ownership = OwnershipBox;
+    type RustType = Arc<dyn ServerCertVerifier>;
```

Comment:
The same comment I mentioned above about fat pointers should be here; or we should put something in the CONTRIBUTING that explains it and link to that.

Reply:
Done.